deepchem/utils/equivariance_utils.py

```diff
@@ -3,6 +3,270 @@
+    """
+
+    def __init__(self):
+        self.leg = {}
+
+    def clear(self) -> None:
+        """Clear cached Legendre polynomial values to save RAM memory."""
+        self.leg = {}
+
+    def get_element(self, d: int, m: int, theta: torch.Tensor,
+                    phi: torch.Tensor) -> torch.Tensor:
+        """
+        Compute a single tesseral spherical harmonic Y_l^m(theta, phi).
+
+        Parameters
+        ----------
+        l : int
+            Degree of the spherical harmonic.
+        m : int
+            Order of the spherical harmonic.
+        theta : torch.Tensor
+            Tensor of polar angles (collatitude) in radians.
+        phi : torch.Tensor
+            Tensor of azimuthal angles (longitude) in radians.
+
+        Returns
+        -------
+        torch.Tensor
+            Tensor of the spherical harmonic values, same shape as `theta`.
+
+        Examples
+        --------
+        >>> theta = torch.tensor([0.0, math.pi / 2])
+        >>> phi = torch.tensor([0.0, math.pi])
+        >>> SphericalHarmonics().get_element(1, 0, theta, phi)
+        tensor([ 4.8860e-01, -2.1357e-08])
+        """
+
+        N = math.sqrt((2 * d + 1) / (4 * math.pi))
+        leg = lpmv(d, abs(m), torch.cos(theta))
+        if m == 0:
+            return N * leg
+        elif m > 0:
+            Y = torch.cos(m * phi) * leg
+        else:
+            Y = torch.sin(abs(m) * phi) * leg
+        N *= math.sqrt(2.0 / pochhammer(d - abs(m) + 1, 2 * abs(m)))
+        Y *= N
+        return Y
+
+    def get(self,
+            d: int,
+            theta: torch.Tensor,
+            phi: torch.Tensor,
+            refresh=True) -> torch.Tensor:
+        """
+        Compute all spherical harmonics of degree l.
+
+        Parameters
+        ----------
+        l : int
+            Degree of the spherical harmonics.
+        theta : torch.Tensor
+            Tensor of polar angles (collatitude) in radians.
+        phi : torch.Tensor
+            Tensor of azimuthal angles (longitude) in radians.
+
+        Returns
+        -------
+        torch.Tensor
+            A tensor of shape [*theta.shape, 2 * l + 1].
+
+        Examples
+        --------
+        >>> theta = torch.tensor([0.0, math.pi / 2])
+        >>> phi = torch.tensor([0.0, math.pi])
+        >>> SphericalHarmonics().get(1, theta, phi)
+        tensor([[-0.0000e+00,  4.8860e-01, -0.0000e+00],
+                [ 4.2715e-08, -2.1357e-08,  4.8860e-01]])
+        """
+        if refresh:
+            self.clear()
+        results = []
+        for m in range(-d, d + 1):
+            results.append(self.get_element(d, m, theta, phi))
+        return torch.stack(results, dim=-1)
+
+
+def irr_repr(order: int,
+             alpha: torch.Tensor,
+             beta: torch.Tensor,
+             gamma: torch.Tensor,
+             dtype: Optional[torch.dtype] = None) -> torch.Tensor:
+    """
+    Compute the irreducible representation of the special orthogonal group SO(3).
```

Comment:
Add a paragraph explaining how the irreducible representation is outputted (its format)